### ISSUE TITLE
Add ability to set aliases (custom names) for uids

### ISSUE THREAD

**Issue by me2beats** [godotengine/godot-proposals#5026]
### Describe the project you are working on

Games

### Describe the problem or limitation you are having in your project

Godot 4 has a cool feture - resources  uids
This allows to get rid of `load(`path/to/resource_file`)`
these paths are not handy because when you change the path (move/rename file), you should update these paths in the code manually.
Also these paths can be quite long.
So now we can use `load("uid://btj0bxdws38cd")` but there are some problems:
- this is even less readable by human, you can't memorize it
- you don't know what is the path of the file
- you can't use code completion

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Being able to set aliases (custom names) for uids would solve the problem.
Uids still exist, you can still use them, but you can also set some more readable names for them.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

uid aliases (custom names):
- should be unique per project
- can be configured in ProjectSetting
- can be stored in project.godot I think
- could be used like `load("uid://my_readable_uid")`
- should be available for code completion

### If this enhancement will not be used often, can it be worked around with a few lines of script?

maybe but not with a few lines. I guess plugins can help.

### Is there a reason why this should be core and not an add-on in the asset library?

I think this would be used very often

**Comment by dalexeev** [comment]
UIDs are human-unfriendly and should not be used in place of file paths. Short human-readable global resource names are a completely separate issue. Also a reminder that you can use relative paths in `preload` and there is a PR adding support for relative paths in `load` (only in GDScript as far as I remember). In addition, in 3.x, the use of `preload` is taken into account in the resource dependency graph, we could automatically change the strings in `preload` when renaming the corresponding resources.

I don't think adding global resource names is necessary. File paths are part of the project's naming conventions. Yes, adding another naming scheme will allow you to move/rename files in some cases without worrying about dependencies. But really, it just postpones the responsibility. If you give a bad global name to a resource and later want to rename it, you still have to check all dependencies using that name and change it to a new one. Same as in the case of using file paths. So in any case, it's better to give good names initially and then you won't have to rename. Otherwise, you need to check all dependencies of this name. For resources and scenes this happens automatically, for scripts you can use the code search.

You can also for example use a singleton (or static class) `Loader` with a dictionary "global resource name -> path/uid", a `load` function using global names instead of paths/uids and implement a plugin that will display a field for editing the global name in the resource inspector.

By the way, I want to say that I am not very happy with UIDs and do not quite understand the purpose of their introduction. As far as I understand there are two:

1. Support for renaming/moving files from external programs. Not sure if this is absolutely necessary. In addition, Git is able to detect renames even with small file changes, and for renames without changing the content, it is enough to use the hash sums of the files.
2. The problem of phantom diffs. It is solved very simply without UID, we can use the same approach as with subresources: instead of a numerical sequential id, use the TypeName + a random sequence of characters. It would be even better if you could manage these names in the editor.

**Comment by me2beats** [comment]
> UIDs are human-unfriendly and should not be used in place of file paths.
Why then there is a brand new syntax load("uid://..") ?
It seems the developers think uids can be used instead of filenames

**Comment by me2beats** [comment]
> You can use relative paths in preload and there is a PR adding support for relative paths in load

Yes, but it is limited by its "relativity". You need to be in a folder a/b to say "get a file "c" that would mean "a/b/c", you can't do this if the script is in folder "a", in this case you have to type "b/c"

**Comment by me2beats** [comment]
> we could automatically change the strings in preload
Sounds interesting but wouldn't it be buggier and less expected?
Will it change all scripts at once?
How would it treat the scripts opened in the script editor which are unsaved (dirty)
Will it correctly create a new undo point etc

**Comment by me2beats** [comment]
> If you give a bad global name to a resource and later want to rename it, you still have to check all dependencies using that name and change it to a new one
Yes but I think it would be much more rare than changing a path itself.

**Comment by me2beats** [comment]
> the purpose of their introduction

Some of the proposes are here I guess
https://github.com/godotengine/godot/issues/15673

> So assets would be loaded via the following pseudo code:
// This will not break if the user moves the material
mesh.material = LoadResource(resource_paths[material_id]);
instead of:
// This will break if the user moves the material
mesh.material = LoadResource(material_path);

**Comment by me2beats** [comment]
Reduz said in Twitter

>For those wondering what this is useful for:
Add-ons can be installed anywhere, like in Unity.
Can move and rename files from a file explorer, version control, or anywhere with the editor closed and they still work.
Keep references to files from code even if moved or renamed (if opening with the uid:// syntax).

**Comment by Delsin-Yu** [comment]
> Support for renaming/moving files from external programs. Not sure if this is absolutely necessary. 

I believe the whole point of the UID is to provide proper support for IDEs, not Git integration.

It's important to note that for members of the Godot Engine team who may not have backgrounds in game development, external tools like IDEs are a common practice. Additionally, since Godot does not offer coding features for the C# language, using an IDE becomes essential.

For developers working on large-scale projects, it's normal to rename files and move them around during refactoring. However, this process can be frustrating without UID file support, as switching back and forth between the Godot Engine and the IDE can be cumbersome.

The IDE itself provides support for renaming and moving related files. For example, when you rename or move `MyScript.cs`, it automatically does the same with `MyScript.uid` (in fact, that's what JetBrains Rider is offering to Unity and Godot); hence, the reference is preserved when you switch back to Godot Editor.
